# Метод DBSCAN

In [1]:
import numpy as np
import pandas as pd

import os

In [2]:
os.chdir("C:/Users/79502/Downloads/Lect_04/Lect_04/задача_2_Economics of Cities/")

In [3]:
df = pd.read_csv("Econom_Cities_data.csv", sep=";", decimal=',', index_col='City')

In [4]:
# Смотрим первые строки, всё ли нормально считалось


df.head()

Work  Price  Salary
City                          
Amsterdam  1714   65.6    49.0
Athens     1792   53.8    30.4
Bogota     2152   37.9    11.5
Bombay     2052   30.3     5.3
Brussels   1708   73.8    50.5

In [5]:
# Данных не очень много, распечатаем всё (вдруг, пригодится для интерпретации результатов)

print(df)

                Work  Price  Salary
City                               
Amsterdam       1714   65.6    49.0
Athens          1792   53.8    30.4
Bogota          2152   37.9    11.5
Bombay          2052   30.3     5.3
Brussels        1708   73.8    50.5
Buenos_Aires    1971   56.1    12.5
Cairo          -9999   37.1 -9999.0
Caracas         2041   61.0    10.9
Chicago         1924   73.9    61.9
Copenhagen      1717   91.3    62.9
Dublin          1759   76.0    41.4
Dusseldorf      1693   78.5    60.2
Frankfurt       1650   74.5    60.4
Geneva          1880   95.9    90.3
Helsinki        1667  113.6    66.6
Hong_Kong       2375   63.8    27.8
Houston         1978   71.9    46.3
Jakarta        -9999   43.6 -9999.0
Johannesburg    1945   51.1    24.0
Kuala_Lumpur    2167   43.5     9.9
Lagos           1786   45.2     2.7
Lisbon          1742   56.2    18.8
London          1737   84.2    46.2
Los_Angeles     2068   79.8    65.2
Luxembourg      1768   71.1    71.1
Madrid          1710   93.8 

In [6]:
from sklearn import preprocessing

# Данные далеко не нормированы, столбцы сильно отличаются. Нужна стандартизация.

stand = preprocessing.StandardScaler()
# вычисляем параметры стандартизации
stand.fit(df)

# преобразуем данные
X = stand.transform(df)

# Получаем объекты X -- матрицы класса numpy.ndarray.
# Преобразуем её в объекты класса pandas.DataFrame.
X = pd.DataFrame(X, index=df.index, columns=df.columns)

In [7]:
# просто посмотрим,что вышло в одной строке
print(X.loc["Zurich"])

Work      0.202983
Price     1.444554
Salary    0.238635
Name: Zurich, dtype: float64


In [32]:
# добавлю эту кластеризацию,пригодится для сравнения

from sklearn.cluster import KMeans
model2 = KMeans(n_clusters=5)
model2.fit(X)
df['KMeans'] = model2.labels_ + 1  # прибавила 1,чтоб начиналось с 1, а не с 0.
df.groupby('KMeans').mean()

Work       Price       Salary   cluster  DB!    DBSCAN
KMeans                                                               
1       1879.933333   69.020000    44.113333  4.600000  0.0  0.000000
2      -9999.000000   40.350000 -9999.000000  2.000000  1.0 -1.000000
3       1760.600000  111.080000    67.500000  4.600000  0.0  1.800000
4       1952.333333   46.746667    12.953333  2.333333  0.0 -0.066667
5       1835.363636   84.790909    56.872727  4.090909  0.0  0.272727

In [33]:
df.groupby('KMeans').size()

KMeans
1    15
2     2
3     5
4    15
5    11
dtype: int64

In [34]:
df['KMeans'].sort_values()

City
Amsterdam         1
Houston           1
Hong_Kong         1
Seoul             1
Frankfurt         1
Singpore          1
Dublin            1
Montreal          1
Luxembourg        1
Caracas           1
Sydney            1
Brussels          1
Tel_Aviv          1
Toronto           1
Chicago           1
Cairo             2
Jakarta           2
Oslo              3
Tokyo             3
Stockholm         3
Zurich            3
Helsinki          3
Lisbon            4
Athens            4
Bogota            4
Bombay            4
Buenos_Aires      4
San_Paulo         4
Rio_de_Janeiro    4
Lagos             4
Nicosia           4
Panama            4
Nairobi           4
Johannesburg      4
Mexico_City       4
Manila            4
Kuala_Lumpur      4
Milan             5
Paris             5
Geneva            5
Dusseldorf        5
Copenhagen        5
Madrid            5
Taipei            5
Vienna            5
London            5
New_York          5
Los_Angeles       5
Name: KMeans, dtype: int32

# Кластеризация методом DBSCAN

In [17]:
from sklearn.cluster import DBSCAN

In [18]:
from scipy.cluster.hierarchy import dendrogram, linkage, fcluster


In [20]:
#  Создадим объект DBSCAN, зададим значения параметров. 
#  Значения 3-х первых параметров совпадают со значениями "по умолчанию"
dbscan_1 = DBSCAN(eps=0.5, metric='euclidean', min_samples=5)


#  Обучим модель  DBSCAN
dbscan_1.fit(X)

#  к какому кластеру относится объект?
# В какой кластер попало наблюдение - Numpy массив номеров кластеров

dbscan_1.labels_


array([ 0,  0,  0,  0,  0,  0, -1,  0,  0,  0,  0,  0,  0,  0, -1,  0,  0,
       -1,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, -1,
        0,  0,  0,  0,  0,  0, -1,  0,  0,  0, -1,  0,  0,  0],
      dtype=int64)

In [21]:
#  В numpy создаем таблицу частот 

#  Не работает. Команда value_counts() работает с таблицами
#  dbscan_1.labels_.value_counts()

#  Нужна замена value_counts() из pandas или замена table из R 
#  Вот вариант замены

unique, counts = np.unique(dbscan_1.labels_, return_counts=True)
print(np.asarray((unique, counts)).T)

[[-1  6]
 [ 0 42]]


In [23]:
dbscan_1 = DBSCAN(eps=2, metric='euclidean', min_samples=1)

dbscan_1.fit(X)

DBSCAN(algorithm='auto', eps=2, leaf_size=30, metric='euclidean',
       metric_params=None, min_samples=1, n_jobs=None, p=None)

In [24]:
#  Таблица частот 

unique, counts = np.unique(dbscan_1.labels_, return_counts=True)
print(np.asarray((unique, counts)).T)

[[ 0 46]
 [ 1  2]]


Вероятно, класс 1 - наши два выброса. Проверим:

In [25]:
df['DB!'] = dbscan_1.labels_
df['DB!'].sort_values()

City
Amsterdam         0
Manila            0
Mexico_City       0
Milan             0
Montreal          0
Nairobi           0
New_York          0
Nicosia           0
Oslo              0
Panama            0
Madrid            0
Paris             0
San_Paulo         0
Seoul             0
Singpore          0
Stockholm         0
Sydney            0
Taipei            0
Tel_Aviv          0
Tokyo             0
Toronto           0
Rio_de_Janeiro    0
Luxembourg        0
Los_Angeles       0
London            0
Athens            0
Bogota            0
Bombay            0
Brussels          0
Buenos_Aires      0
Caracas           0
Chicago           0
Copenhagen        0
Dublin            0
Vienna            0
Dusseldorf        0
Geneva            0
Helsinki          0
Hong_Kong         0
Houston           0
Johannesburg      0
Kuala_Lumpur      0
Lagos             0
Lisbon            0
Frankfurt         0
Zurich            0
Cairo             1
Jakarta           1
Name: DB!, dtype: int64

Действительно! Так, уже есть кластеризация на выбросы и невыбросы. Значения в них действительно похожи. И если не знать, что там -9999, можно подумать, что это точки в одном классе. Идём дальше.

In [27]:
dbscan_1 = DBSCAN(eps=0.3, metric='euclidean', min_samples=2)

dbscan_1.fit(X)
unique, counts = np.unique(dbscan_1.labels_, return_counts=True)
print(np.asarray((unique, counts)).T)

[[-1  3]
 [ 0 37]
 [ 1  4]
 [ 2  4]]


Надо посмотреть, что вышло.

In [28]:
df['DBSCAN'] = dbscan_1.labels_
df['DBSCAN'].sort_values()

City
Bombay           -1
Jakarta          -1
Cairo            -1
Amsterdam         0
Manila            0
Mexico_City       0
Milan             0
Montreal          0
Nairobi           0
New_York          0
Nicosia           0
Panama            0
Paris             0
Rio_de_Janeiro    0
San_Paulo         0
Seoul             0
Singpore          0
Sydney            0
Taipei            0
Tel_Aviv          0
Toronto           0
Luxembourg        0
Vienna            0
Los_Angeles       0
Lisbon            0
Athens            0
Bogota            0
Brussels          0
Buenos_Aires      0
Caracas           0
Chicago           0
Dublin            0
London            0
Frankfurt         0
Dusseldorf        0
Hong_Kong         0
Houston           0
Johannesburg      0
Kuala_Lumpur      0
Lagos             0
Geneva            1
Copenhagen        1
Madrid            1
Zurich            1
Helsinki          2
Stockholm         2
Tokyo             2
Oslo              2
Name: DBSCAN, dtype: int64

Получилось три кластера:
кластеры 2 и 1 - тот, что раньше был высоким уровнем жизни, оствльное - вместе, плюс - класс из выбросов перекочевал в выбросы, к которым зачем-то присоединился Каир. Странно всё. Пробуем еще.

In [49]:
dbscan_1 = DBSCAN(eps=0.15, metric='euclidean', min_samples=3)

dbscan_1.fit(X)
unique, counts = np.unique(dbscan_1.labels_, return_counts=True)
print(np.asarray((unique, counts)).T)

[[-1  7]
 [ 0 21]
 [ 1 13]
 [ 2  4]
 [ 3  3]]


In [57]:
df['DBSCAN'] = dbscan_1.labels_
df['DBSCAN'].sort_values()

City
Zurich           -1
Bogota           -1
Bombay           -1
Manila           -1
Cairo            -1
Caracas          -1
Jakarta          -1
Vienna            0
Luxembourg        0
Milan             0
Montreal          0
Singpore          0
Paris             0
Sydney            0
Taipei            0
Tel_Aviv          0
Toronto           0
New_York          0
London            0
Amsterdam         0
Los_Angeles       0
Brussels          0
Houston           0
Hong_Kong         0
Chicago           0
Frankfurt         0
Dusseldorf        0
Dublin            0
Athens            1
Buenos_Aires      1
Seoul             1
San_Paulo         1
Rio_de_Janeiro    1
Panama            1
Lagos             1
Nicosia           1
Nairobi           1
Mexico_City       1
Johannesburg      1
Lisbon            1
Kuala_Lumpur      1
Stockholm         2
Helsinki          2
Tokyo             2
Oslo              2
Geneva            3
Copenhagen        3
Madrid            3
Name: DBSCAN, dtype: int64

Интересно! Посмотрим средние значения.

In [56]:
df.groupby('DBSCAN').mean()

Work       Price       Salary    KMeans
DBSCAN                                                
-1     -1373.857143   49.985714 -2838.042857  2.857143
 0      1866.571429   74.652381    49.547619  2.523810
 1      1896.538462   50.100000    15.861538  3.769231
 2      1733.750000  113.850000    59.375000  3.000000
 3      1769.000000   93.666667    67.733333  5.000000

-1 - выбросы, не смотрим, там всё подряд.

Кластер 2 - это высокие цены, сравнительно маленькое время работы, достойная (но не самая высокая!) зарплата. Это страны с высоким уровнем жизни, и, посмею предположить, с какой-то хорошей социальной поддержкой (зарплата не самая большая, а цены
просто атомные!!!!) И уже даже по этой таблице видно, что кластер полностью совпал со кластером 3 из метода к-средних.

Кластер 3 - тут зарплата даже выше!!! А цены тут пониже. Пусть будет: города, в которых хорошо работать, но дорого жить.

Кластер 0 - среднячки. Цены намного ниже, чем в прошлых кластерах. Зарплата тоже ниже, но не настолько. Нормальные такие города.

Кластер 1 - нищие города. Очень низкая зарплата, число рабочих часов - максимальное! Цены минимальные, но по сравнению с зарплатой они не так сильно снизились. Печальное место.

Сравним с  K-Means:

In [58]:
tab = pd.crosstab(df['DBSCAN'], df['KMeans'])

print(tab)

KMeans   1  2  3   4  5
DBSCAN                 
-1       1  2  1   3  0
 0      13  0  0   0  8
 1       1  0  0  12  0
 2       0  0  4   0  0
 3       0  0  0   0  3


Очень большое сходство! Правда, некоторые города попали в выбросы. Два класса(1 и 5 -> 0) слились, а кусок оставшегося отправился в отдельный класс. 

In [80]:
dbscan_1 = DBSCAN(eps=0.1575, metric='euclidean', min_samples=3)

dbscan_1.fit(X)
unique, counts = np.unique(dbscan_1.labels_, return_counts=True)
print(np.asarray((unique, counts)).T)

[[-1  6]
 [ 0 21]
 [ 1 14]
 [ 2  4]
 [ 3  3]]


In [78]:
df['DBSCAN'] = dbscan_1.labels_
df['DBSCAN'].sort_values()

City
Zurich           -1
Manila           -1
Cairo            -1
Jakarta          -1
Bogota           -1
Bombay           -1
Tel_Aviv          0
Taipei            0
Sydney            0
London            0
Paris             0
New_York          0
Montreal          0
Milan             0
Toronto           0
Luxembourg        0
Vienna            0
Singpore          0
Amsterdam         0
Los_Angeles       0
Brussels          0
Houston           0
Hong_Kong         0
Chicago           0
Frankfurt         0
Dusseldorf        0
Dublin            0
Athens            1
Buenos_Aires      1
Caracas           1
Seoul             1
San_Paulo         1
Rio_de_Janeiro    1
Panama            1
Kuala_Lumpur      1
Nicosia           1
Nairobi           1
Mexico_City       1
Lisbon            1
Lagos             1
Johannesburg      1
Stockholm         2
Helsinki          2
Tokyo             2
Oslo              2
Copenhagen        3
Geneva            3
Madrid            3
Name: DBSCAN, dtype: int64

Один город перекочевал из выбросов в среднячки, но особо изменений нет.

In [86]:
dbscan_1 = DBSCAN(eps=0.16, metric='euclidean', min_samples=2)

dbscan_1.fit(X)
unique, counts = np.unique(dbscan_1.labels_, return_counts=True)
print(np.asarray((unique, counts)).T)

[[-1  4]
 [ 0 35]
 [ 1  2]
 [ 2  3]
 [ 3  4]]


Такое же количество кластеров. Но очень неравномерное разбиение! Зато меньше выбросов! Посмотрим:

In [89]:
df['DBSCAN'] = dbscan_1.labels_
df['DBSCAN'].sort_values()

City
Zurich           -1
Bombay           -1
Jakarta          -1
Cairo            -1
Mexico_City       0
Milan             0
Montreal          0
Nairobi           0
New_York          0
Nicosia           0
Panama            0
Rio_de_Janeiro    0
Luxembourg        0
San_Paulo         0
Seoul             0
Singpore          0
Sydney            0
Taipei            0
Tel_Aviv          0
Toronto           0
Paris             0
Vienna            0
Amsterdam         0
Lisbon            0
Athens            0
Brussels          0
Buenos_Aires      0
Caracas           0
Chicago           0
Dublin            0
London            0
Frankfurt         0
Dusseldorf        0
Hong_Kong         0
Houston           0
Johannesburg      0
Kuala_Lumpur      0
Lagos             0
Los_Angeles       0
Manila            1
Bogota            1
Copenhagen        2
Madrid            2
Geneva            2
Oslo              3
Helsinki          3
Stockholm         3
Tokyo             3
Name: DBSCAN, dtype: int64

In [90]:
df.groupby('DBSCAN').mean()

Work       Price       Salary    KMeans
DBSCAN                                                
-1     -4019.500000   52.750000 -4973.175000  2.750000
 0      1882.685714   65.142857    35.931429  2.942857
 1      2210.000000   38.950000     7.750000  4.000000
 2      1769.000000   93.666667    67.733333  5.000000
 3      1733.750000  113.850000    59.375000  3.000000

Появился супер-бедный класс. Чем-то похоже на прошлое разбиение. Но если присмотреться - Найроби и Нью-Йорк в одном классе!!!


In [97]:
print(df.loc["Nairobi"])
print(df.loc["New_York"])

Work      1958.0
Price       45.0
Salary       5.8
KMeans       4.0
DBSCAN       0.0
Name: Nairobi, dtype: float64
Work      1942.0
Price       83.3
Salary      65.8
KMeans       5.0
DBSCAN       0.0
Name: New_York, dtype: float64


* Nairobi         1958   45.0     5.8
* New_York        1942   83.3    65.8
Как сильно отличается зарплата! А цены? Это разбиение мне не нравится! Не буду гнаться за минимизацией выбросов. Остановимся на предыдущем разбиении! Оно было самое адекватное.

Мой вывод: этом метод кластеризации понравился мне меньше всего: очень сложно правильно повлиять на результат, разбиения самые непонятные. Может, для ленточных кластеров он хорош, но для обычной кластеризации его намного сложнее настраивать. 